File: C#/Solutions/Problem011/Trie.cs
﻿using System.Collections.Generic;
using System.Linq;

namespace DailyCodingProblems.Solutions.Problem011
{
    /// <summary>
    /// Prefix tree. https://en.wikipedia.org/wiki/Trie
    /// </summary>
    public class Trie
    {
        private readonly TrieNode _head = new TrieNode();

        public void Add(string word)
        {
            Add(_head, word);
        }

        private static void Add(TrieNode node, string subString)
        {
            var currentString = string.Empty;

            while (true)
            {
                if (subString.Length == 0)
                {
                    break;
                }

                var prefixChar = subString[0];
                if (!node.SubNodes.ContainsKey(prefixChar))
                {
                    node.SubNodes.Add(prefixChar, new TrieNode(currentString + prefixChar));
                }

                var remainingSubString = subString[1..];

                if (remainingSubString.Length == 0)
                {
                    node.SubNodes[prefixChar].IsWord = true;
                    break;
                }

                node = node.SubNodes[prefixChar];
                subString = remainingSubString;
                currentString += prefixChar;
            }
        }

        public IReadOnlyCollection<string> Search(string searchString)
        {
            var node = _head;
            foreach (var searchChar in searchString)
            {
                if (!node.SubNodes.ContainsKey(searchChar))
                {
                    return new string[0];
                }
                node = node.SubNodes[searchChar];
            }

            return Search(node);
        }

        private static IReadOnlyCollection<string> Search(TrieNode node)
        {
            var returnList = new List<string>();
            if (node.IsWord)
            {
                returnList.Add(node.Word);
            }

            returnList.AddRange(node.SubNodes.SelectMany(subNode => Search(subNode.Value)));

            return returnList;
        }

        /// <summary>
        /// Node in a <see cref="Trie"/>
        /// </summary>
        public class TrieNode
        {
            public IDictionary<char, TrieNode> SubNodes { get; }
            public string Word { get; }
            public bool IsWord { get; set; }

            public TrieNode(string word = "")
            {
                SubNodes = new Dictionary<char, TrieNode>();
                Word = word;
                IsWord = false;
            }
        }
    }
}

File: Java/Solutions/Problem011/Trie.java
package Problem011;

import java.util.*;

/**
 * Prefix tree. https://en.wikipedia.org/wiki/Trie
 */
public class Trie {
    private final TrieNode _head = new TrieNode("");

    public void add(String word) {
        add(_head, word);
    }

    private void add(TrieNode node, String subString) {
        var currentString = "";
        while(true) {
            if (subString.length() == 0) {
                break;
            }

            var prefixChar = subString.charAt(0);
            if (!node.getSubNodes().containsKey(prefixChar)) {
                node.getSubNodes().put(prefixChar, new TrieNode(currentString + prefixChar));
            }

            var remainingSubString = subString.substring(1);
            if (remainingSubString.length() == 0) {
                node.getSubNodes().get(prefixChar).setWord(true);
                break;
            }

            node = node.getSubNodes().get(prefixChar);
            subString = remainingSubString;
            currentString += prefixChar;
        }
    }

    public List<String> search(String searchString) {
        var node = _head;
        var charArray = new char[searchString.length()];
        searchString.getChars(0, searchString.length(), charArray, 0);

        for (var searchChar:
             charArray) {
            if (!node.getSubNodes().containsKey(searchChar)) {
                return Collections.emptyList();
            }
            node = node.getSubNodes().get(searchChar);
        }
        return search(node);
    }

    private List<String> search(TrieNode node) {
        var returnList = new ArrayList<String>();
        if (node.isWord()) {
            returnList.add(node.getWord());
        }

        for (var subNode:
             node.getSubNodes().entrySet()) {
            for (var result:
                 search(subNode.getValue())) {
                returnList.add(result);
            }
        }
        return returnList;
    }

    /**
     * Node in a Trie
     */
    protected class TrieNode {
        private final Map<Character, TrieNode> subNodes;
        private final String word;
        private boolean isWord;

        public TrieNode(String word) {
            this.subNodes = new HashMap<>();
            this.word = word;
            this.isWord = false;
        }

        public Map<Character, TrieNode> getSubNodes() {
            return subNodes;
        }

        public String getWord() {
            return word;
        }

        public boolean isWord() {
            return isWord;
        }

        public void setWord(boolean word) {
            isWord = word;
        }
    }
}
